SimplenoteTests/NSMutableAttributedStringStylingTests.swift

```diff
@@ -94,5 +104,22 @@ class NSMutableAttributedStringStylingTests: XCTestCase {
         let matches = regex.matches(in: string, options: [], range: string.fullRange)
 
         XCTAssertEqual(matches.count, 1)
+        XCTAssertTrue(verifyEveryMatch(in: matches, contains: numberOfExpectedRangesPerMatch))
+    }
+}
+
+
+// MARK: - Internal Helpers
+//
+private extension NSMutableAttributedStringStylingTests {
+
+    /// Verifies that all of the TextCheckingResults contain the specified number of ranges
+    ///
+    func verifyEveryMatch(in matches: [NSTextCheckingResult], contains numberOfRanges: Int) -> Bool {
+        for match in matches where match.numberOfRanges != numberOfRanges {
+            return false
+        }
+
+        return true
```

Comment:
this is novel. did you try `.contains` already?

```swift
func verifyEveryMatch(in matches: [NSTextCheckingResult], contains numberOfRanges: Int) -> Bool {
  return matches.contains { match in match.numberOfRanges != numberOfRanges }
}
```

what which point, is this function necessary?

```swift
XCTAssertEqual(matches.count, 1)
XCTAssertTrue(matches.contains { match in match.numberOfRanges != numberOfExpectedRangesPerMatch } )
```

or alternatively in the positive…

```swift
XCTAssertEqual(matches.count, 1)
XCTAssertTrue(matches.allSatisfy({ $0.numberOfRanges == numberOfExpectedRangesPerMatch }))
```

Reply:
Hey there sir!! Yeah I'm familiar with those APIs, **but** personally I try to avoid them.

**Why:** Because you're passing over a closure (anonymous function) which will be executed N times. It's just all about performance.

I know it's a tiny difference, but in the long run it truly adds up.

Thanks for the heads up!!!

Reply:
> I know it's a tiny difference, but in the long run it truly adds up.

But does it?
Did you measure it?

I mean, these aren't some rando library functions - they're part of the `NSArray` class in the standard library. Also, languages like Swift are going to have optimizations for closures, especially ones as trivial as this. It might not actually pass any closures after it compiles, and I'm pretty sure in either case it's significantly faster than the time it takes to run the RegExp.

So, it's your code - you make the call. I wanted to ask in case you weren't aware of these, whose purpose is to tersely communicate the intent of the check "does everything follow this rule?" vs. the _steps_ we're taking "call this function, iterate over all the items, and if any one doesn't follow the rule then return false" (which themselves could contain bugs but for which no explanation is readily at hand). And now the thought of avoiding them for performance seems dubious at best.

In other words, it would seem just as silly to use them for performance as to avoid them but do as you like. I'm sure the code already has a bunch of its own idiosyncrasies like this in it and there's value in consistency.

Reply:
Hey D!! let's take this one to Slack, since this PR fixes a quite annoying bug, and we've got a release scheduled next monday.

Thaaaaankyoouu!!